src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentFieldIndexProvider.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json.Linq;
+using OrchardCore.ContentFields.Fields;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Metadata;
+using OrchardCore.Data;
+using YesSql.Indexes;
+
+namespace OrchardCore.ContentFields.Indexing
+{
+    // Remark: 
+
+    public class ContentFieldIndex : MapIndex
+    {
+        public string ContentType { get; set; }
+        public string ContentPart { get; set; }
+        public string ContentField { get; set; }
+    }
+
+    public class ContentFieldIndexProvider : IndexProvider<ContentItem>, IScopedIndexProvider
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly HashSet<string> _ignoredTypes = new HashSet<string>();
+        private IContentDefinitionManager _contentDefinitionManager;
+
+        public ContentFieldIndexProvider(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        public override void Describe(DescribeContext<ContentItem> context)
+        {
+            context.For<ContentFieldIndex>()
+                .Map(contentItem =>
+                {
+                    if (!contentItem.IsPublished())
+                    {
+                        return null;
+                    }
+
+                    // Can we safely ignore this content item?
+                    if (_ignoredTypes.Contains(contentItem.ContentType))
+                    {
+                        return null;
+                    }
+
+                    // Lazy initialization because of ISession cyclic dependency
+                    _contentDefinitionManager = _contentDefinitionManager ?? _serviceProvider.GetRequiredService<IContentDefinitionManager>();
+
+                    // Search for Text fields
+                    var fieldDefinitions = _contentDefinitionManager
+                        .GetTypeDefinition(contentItem.ContentType)
+                        .Parts.SelectMany(x => x.PartDefinition.Fields)
+                        .ToArray();
+
+                    var results = new List<ContentFieldIndex>();
+
+                    foreach (var fieldDefinition in fieldDefinitions)
+                    {
+                        var jPart = (JObject)contentItem.Content[fieldDefinition.PartDefinition.Name];
+
+                        if (jPart == null)
+                        {
+                            continue;
+                        }
+
+                        var jField = (JObject)jPart[fieldDefinition.Name];
+
+                        if (jField == null)
+                        {
+                            continue;
+                        }
+
+                        var field = jField.ToObject<TextField>();
+
+                        results.Add(new ContentFieldIndex
```

Comment:
Create an absctract `CreateIndex` that returns a generic type defined on this class. This way the sub-classes can only implement this new method that return the index object only. Lots of code to remove from all providers then. If not clear ping me and we'll pair program.